content/reference/routing.md

```diff
@@ -0,0 +1,406 @@
+
+function view () {
+  return html`
+    <body>
+      <a href="/uh-oh">
+        Click Click Click
+      </a>
+    </body>
+  `
+}
+
+function notFound () {
+  return html`
+    <body>
+      <a href="/">
+        Route not found. Navigate back.
+      </a>
+    </body>
+  `
+}
+```
+
+1. We define our base view on route `/`. This is the first route that's loaded
+   when a person visits our site. It contains a single anchor tag that points
+   to `/uh-oh`, which is a route that doesn't exist.
+2. It's good practice to define a fallback route statically as `/404`. This
+   helps with debugging, and is often treated specially when deploying to
+   production.
+3. We define our fallback route as `*`. The asterix symbol is pronounced
+   "glob". Our glob route will now handle all routes that didn't match
+   anything.
+4. We mount the application on the DOM. If someone now clicks the link that's
+   rendered in `/`, it will be handled by the fallback route.
+
+## Dynamic routing
+Sometimes there will be pages that have the same layout, but different data.
+For example user pages, or blog entries. This requires _dynamic routing_. In
+Choo we have two types of syntax for dynamic routing.
+
+### Params
+Params are declared with the `:` syntax. For example `/foo/:bar`. This means
+that the `/foo` part is static, and `:bar` can be any value, up until the next
+slash (`/`).
+
+The value from a param, is exposed in Choo as `state.params`. So say we have
+the route `/foo/:bar`, and we navigate to `/foo/beep`, the value of
+`state.params.bar` will be `'beep'`.
+
+```js
+var html = require('choo/html')
+var choo = require('choo')
+
+var app = choo()
+app.route('/', placeholder)
+app.route('/:user', placeholder)
+app.route('/:user/:repo', placeholder)
+app.route('/:user/settings', placeholder)
+app.route('/404', placeholder)
+
+function placeholder (state) {
+  console.log(state.params)
+  return html`<body>placeholder</body>`
+}
+```
+
+### Wildcards
+In most cases params are the answer for dynamic routing. They're named, readily
+available, and easy to traverse. However, they can't cover cases where the
+amount of slashes in route will be unknown.
+
+Take for example GitHub's code view. To navigate to Choo's `html/raw` API, the
+route is:
+[github.com/choojs/choo/blob/master/html/raw.js](https://github.com/choojs/choo/blob/master/html/raw.js) .
+In this case, all parts of the route after `master/` are unknown. This means
+that `params` can't match the route correctly, because there's an unknown
+amount of slashes after `master/`. This is what wildcards are for.
+
+If we were building GitHub's code view with Choo, we could express the route
+as: `/:user/:repo/blob/:branch/*`. The value of `state.params.wildcard` would
+then be: `'/html/raw.js'`.
+
+Try and use wildcards sparingly. They're the most powerful tool in the routing
+toolbox, which means that if you're not careful you might end up reimplementing
+a router on top of it yourself.
+
```

Comment:
this is missing a section about query parameters?

Reply:
done!